optaplanner-core/src/main/java/org/optaplanner/core/impl/phase/loop/DefaultLoopPhase.java

```diff
@@ -0,0 +1,164 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.phase.loop;
+
+import java.util.List;
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.impl.phase.AbstractPhase;
+import org.optaplanner.core.impl.phase.PhaseCounter;
+import org.optaplanner.core.impl.phase.event.PhaseLifecycleListener;
+import org.optaplanner.core.impl.phase.event.PhaseLifecycleSupport;
+import org.optaplanner.core.impl.phase.loop.scope.LoopPhaseScope;
+import org.optaplanner.core.impl.phase.loop.scope.LoopStepScope;
+import org.optaplanner.core.impl.solver.recaller.BestSolutionRecaller;
+import org.optaplanner.core.impl.solver.scope.SolverScope;
+import org.optaplanner.core.impl.solver.termination.Termination;
+
+/**
+ * Default implementation of {@link LoopPhase}.
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ */
+public class DefaultLoopPhase<Solution_> extends AbstractPhase<Solution_> implements LoopPhase<Solution_> {
+
+    private final List<AbstractPhase<Solution_>> phaseList;
+
+    public DefaultLoopPhase(PhaseCounter<Solution_> phaseCounter, String logIndentation,
+            BestSolutionRecaller<Solution_> bestSolutionRecaller, Termination<Solution_> termination,
+            List<AbstractPhase<Solution_>> phaseList) {
+        super(phaseCounter, logIndentation, bestSolutionRecaller, termination);
+        this.phaseList = phaseList;
+    }
+
+    @Override
+    public String getPhaseTypeString() {
+        return "Loop";
+    }
+
+    // ************************************************************************
+    // Worker methods
+    // ************************************************************************
+
+    @Override
+    public void solve(SolverScope<Solution_> solverScope) {
+        LoopPhaseScope<Solution_> phaseScope = new LoopPhaseScope<>(solverScope);
+        phaseStarted(phaseScope);
+
+        boolean terminated = false;
+        do { // Loop the nested phases until termination is reached.
+            for (AbstractPhase<Solution_> phase : phaseList) {
+                LoopStepScope<Solution_> stepScope = new LoopStepScope<>(phaseScope);
+                stepStarted(stepScope);
+                doStep(solverScope, stepScope, phase);
+                stepEnded(stepScope);
+                phaseScope.setLastCompletedStepScope(stepScope);
+
+                solverScope.checkYielding();
+                terminated = termination.isPhaseTerminated(phaseScope);
+                if (terminated) { // The termination condition may be met inbetween the nested phases.
+                    break;
+                }
+            }
+        } while (!terminated);
```

Comment:
All other phases do a normal loop, not a do-while loop. They check the termination condition before looping even.

For example DefaultLocalSearchPhase:
```
        while (!termination.isPhaseTerminated(phaseScope)) {
˘˘˘